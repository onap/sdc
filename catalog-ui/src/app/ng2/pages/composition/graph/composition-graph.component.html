<div class="sdc-composition-graph-wrapper {{zoneTagMode}}"
     [ngClass]="{'with-sidebar': withSidebar$ | async, 'view-only':isViewOnly$ | async}">
</div>

<div class="sdc-composition-menu" [ngClass]="{'with-sidebar': withSidebar$ | async}">

    <service-path-selector
            *ngIf="topologyTemplate.isService() && compositionService.forwardingPaths"
            [drawPath]="drawPathOnCy"
            [deletePaths]="deletePathsOnCy"
            [selectedPathId]="selectedPathId">
    </service-path-selector>

    <canvas-search *ngIf="componentInstanceNames" class="composition-search"
                   [placeholder]="'Type to search'"
                   [data]="componentInstanceNames"
                   (searchChanged)="getAutoCompleteValues($event)"
                   (searchButtonClicked)="highlightSearchMatches($event)">
    </canvas-search>

    <!--<service-path class="zoom-icons"-->
                  <!--*ngIf="!(isViewOnly$ | async) && topologyTemplate.isService()"-->
                  <!--[service]="topologyTemplate"-->
                  <!--[onCreate]="createOrUpdateServicePath">-->
    <!--</service-path>-->

    <svg-icon  *ngIf="!(isViewOnly$ | async) && topologyTemplate.isService()" class="zoom-icons" [mode]="'primary2'" [size]="'medium'" [backgroundShape]="'rectangle'"
               [backgroundColor]="'silver'" [name]="'browse'" [clickable]="true" [testId]="'pathsMenuBtn'"
               (click)="openServicePathMenu($event)"></svg-icon>
    <svg-icon class="zoom-icons" [mode]="'primary2'" [size]="'medium'" [backgroundShape]="'rectangle'"
              [backgroundColor]="'silver'" [name]="'expand-o'" [clickable]="true"
              (click)="zoomAllWithoutSidebar()"></svg-icon>
    <svg-icon class="zoom-icons" [mode]="'primary2'" [size]="'medium'" [backgroundShape]="'rectangle'"
              [backgroundColor]="'silver'" [name]="'plus'" [clickable]="true"
              (click)="zoom(true)"></svg-icon>
    <svg-icon class="zoom-icons" [mode]="'primary2'" [size]="'medium'" [backgroundShape]="'rectangle'"
              [backgroundColor]="'silver'" [name]="'minus'" [clickable]="true"
              (click)="zoom(false)"></svg-icon>
</div>

<div class="sdc-canvas-zones__wrapper {{zoneTagMode}}" [ngClass]="{'with-sidebar': withSidebar$ | async}">
    <zone-container *ngFor="let zone of zones" [title]="zone.title" [type]="zone.type" [count]="zone.instances.length"
                    [visible]="zone.visible" [minimized]="zone.minimized" (minimize)="zoneMinimizeToggle(zone.type)"
                    (backgroundClick)="zoneBackgroundClicked()">
        <zone-instance *ngFor="let instance of zone.instances" [hidden]="instance.hidden"
                       [zoneInstance]="instance" [defaultIconText]="zone.defaultIconText"
                       [isActive]="activeZoneInstance == instance"
                       [activeInstanceMode]="activeZoneInstance && activeZoneInstance.mode"
                       [isViewOnly]="isViewOnly$ | async"
                       [forceSave]="instance.forceSave"
                       (modeChange)="zoneInstanceModeChanged($event.newMode, $event.instance, zone.type)"
                       (tagHandleClick)="zoneInstanceTagged($event)"
                       (assignmentSaveStart)="zoneAssignmentSaveStart()"
                       (assignmentSaveComplete)="zoneAssignmentSaveComplete($event)">
        </zone-instance>
    </zone-container>
</div>