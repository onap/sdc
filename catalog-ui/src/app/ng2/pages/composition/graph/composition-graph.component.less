:host(composition-graph) {
  flex: 1;
  padding-top: 53px;
}

.composition {
  .custom-modal {
    /* Hack solution to hide canvas tooltips under modals */
    z-index: 20000 !important;
  }
}

.sdc-composition-graph-wrapper {
  height: 100%;
  width: 100%;

  &.with-sidebar {
    width: calc(~'100% - 300px');
  }
}

.view-only {
  background-color: rgb(248, 248, 248);
}

.sdc-canvas-zones__wrapper {
  position: absolute;
  bottom: 10px;
  right: 12px;
  display: flex;
  transition: right 0.2s;

  &.with-sidebar {
    right: 310px;
  }

  ng2-zone-container {
    display: flex;
    margin-left: 10px;
  }
}

.group-tagging {
  cursor: url("../../../../../assets/styles/images/canvas-tagging-icons/group_1.svg"), pointer;
}

.group-tagging-hover {
  cursor: url("../../../../../assets/styles/images/canvas-tagging-icons/group_2.svg"), pointer;
}

.policy-tagging {
  cursor: url("../../../../../assets/styles/images/canvas-tagging-icons/policy_1.svg"), pointer;
}

.policy-tagging-hover {
  cursor: url("../../../../../assets/styles/images/canvas-tagging-icons/policy_2.svg"), pointer;
}

//Canvas menu
.sdc-composition-menu {
  position: absolute;
  right: 18px;
  top: 53px;
  transition: right 0.2s;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  margin-right: 10px;
  pointer-events: none;

  & > * {
    pointer-events: all;
  }

  &.with-sidebar {
    right: 320px;
  }

  .composition-search {
    margin-top: 12px;
  }

  .zoom-icons {
    border: solid 1px #d2d2d2;
    border-radius: 2px;
    box-shadow: 0px 2px 3.88px 0.12px rgba(0, 0, 0, 0.29);
    margin-top: 10px;

    /deep/ .svg-icon {
      box-sizing: content-box;
    }
  }
}