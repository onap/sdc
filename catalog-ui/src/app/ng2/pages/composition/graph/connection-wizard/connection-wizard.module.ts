import {ToNodeStepComponent} from "./to-node-step/to-node-step.component";
import {NgModule} from "@angular/core";
import {FromNodeStepComponent} from "./from-node-step/from-node-step.component";
import {PropertiesStepComponent} from "./properties-step/properties-step.component";
import {ConnectionWizardService} from "./connection-wizard.service";
import {SelectRequirementOrCapabilityModule} from "../../../../components/logic/select-requirement-or-capability/select-requirement-or-capability.module";
import {PropertyTableModule} from "../../../../components/logic/properties-table/property-table.module";
import {FormElementsModule} from "../../../../components/ui/form-components/form-elements.module";
import {ConnectionWizardHeaderComponent} from "./connection-wizard-header/connection-wizard-header.component";
import {ConnectionPropertiesViewComponent} from "./connection-properties-view/connection-properties-view.component";
import {BrowserModule} from "@angular/platform-browser";

@NgModule({
    declarations: [
        FromNodeStepComponent,
        ToNodeStepComponent,
        PropertiesStepComponent,
        ConnectionWizardHeaderComponent,
        ConnectionPropertiesViewComponent
    ],
    imports: [
        FormElementsModule,
        PropertyTableModule,
        SelectRequirementOrCapabilityModule,
        BrowserModule
    ],
    exports: [
        FromNodeStepComponent,
        ToNodeStepComponent,
        PropertiesStepComponent,
        ConnectionWizardHeaderComponent,
        ConnectionPropertiesViewComponent
    ],
    entryComponents: [FromNodeStepComponent,
        ToNodeStepComponent,
        PropertiesStepComponent,
        ConnectionWizardHeaderComponent,
        ConnectionPropertiesViewComponent
    ],
    providers: [ConnectionWizardService]
})
export class ConnectionWizardModule {
}