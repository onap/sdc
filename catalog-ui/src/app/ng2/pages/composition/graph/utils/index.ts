/**
 * Created by ob0695 on 6/3/2018.
 */
// export * from './composition-graph-general-utils';
// export * from './composition-graph-links-utils';
// export * from './composition-graph-nodes-utils';
// export * from './composition-graph-palette-utils';
// export * from './composition-graph-service-path-utils';
// export * from './composition-graph-zone-utils';


import {CompositionGraphGeneralUtils} from './composition-graph-general-utils';
import {CompositionGraphNodesUtils} from './composition-graph-nodes-utils';
import {MatchCapabilitiesRequirementsUtils} from './match-capability-requirement-utils'
import {CompositionGraphPaletteUtils} from './composition-graph-palette-utils';
import {CompositionGraphZoneUtils} from './composition-graph-zone-utils';
import {ServicePathGraphUtils} from './composition-graph-service-path-utils';
import {CompositionGraphLinkUtils} from "./composition-graph-links-utils";


export {
    CompositionGraphGeneralUtils,
    CompositionGraphLinkUtils,
    CompositionGraphNodesUtils,
    MatchCapabilitiesRequirementsUtils,
    CompositionGraphPaletteUtils,
    CompositionGraphZoneUtils,
    ServicePathGraphUtils
};