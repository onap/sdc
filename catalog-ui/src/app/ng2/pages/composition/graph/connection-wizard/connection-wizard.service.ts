import * as _ from "lodash";
import {ConnectRelationModel} from "app/models/graph/connectRelationModel";
import {Injectable} from "@angular/core";
import { Requirement, Capability} from "app/models";
import {Dictionary} from "lodash";
import {Match, Component, PropertyFEModel} from "app/models";
import {Store} from "@ngxs/store";
import {WorkspaceService} from "../../../workspace/workspace.service";

@Injectable()
export class ConnectionWizardService {
    
    connectRelationModel:ConnectRelationModel;
    selectedMatch:Match;
    changedCapabilityProperties:PropertyFEModel[];


    constructor(private workspaceService: WorkspaceService) {
        this.changedCapabilityProperties = [];

    }

    public setRelationMenuDirectiveObj = (connectRelationModel:ConnectRelationModel) => {
        this.connectRelationModel = connectRelationModel;
        // this.selectedCapability = rel
    }

    getOptionalRequirementsByInstanceUniqueId = (isFromTo: boolean, matchWith?:Capability): Dictionary<Requirement[]> => {
       let requirements: Array<Requirement> = [];
        _.forEach(this.connectRelationModel.possibleRelations, (match: Match) => {
            if(!matchWith || match.capability.uniqueId == matchWith.uniqueId){
                if(match.isFromTo == isFromTo){
                    requirements.push(match.requirement);
                }
            }
        });
        requirements = _.uniqBy(requirements, (req:Requirement)=>{
            return req.ownerId + req.uniqueId + req.name;
        });
        return _.groupBy(requirements, 'capability');
    }

    getOptionalCapabilitiesByInstanceUniqueId = (isFromTo: boolean, matchWith?:Requirement): Dictionary<Capability[]> => {
        let capabilities: Array<Capability> = [];
        _.forEach(this.connectRelationModel.possibleRelations, (match: Match) => {
            if(!matchWith || match.requirement.uniqueId == matchWith.uniqueId){
                if(match.isFromTo == isFromTo){
                    capabilities.push(match.capability);
                }
            }
        });
        capabilities = _.uniqBy(capabilities, (cap:Capability)=>{
            return cap.ownerId + cap.uniqueId + cap.name;
        });
        return _.groupBy(capabilities, 'type');
    }
}

