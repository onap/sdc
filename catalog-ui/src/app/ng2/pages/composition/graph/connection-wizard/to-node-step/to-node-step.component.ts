import {Component, forwardRef, Inject} from '@angular/core';
import {IStepComponent} from "app/models"
import {Dictionary} from "lodash";
import {ConnectionWizardService} from "../connection-wizard.service";
import {Match} from "app/models/graph/match-relation";
import {Requirement} from "app/models/requirement";
import {Capability} from "app/models/capability";
import {PropertyModel} from "app/models/properties";

@Component({
    selector: 'to-node-step',
    templateUrl: './to-node-step.component.html'
})

export class ToNodeStepComponent implements IStepComponent{

    displayRequirementsOrCapabilities:string; //get 'Requirement' or 'Capability'
    optionalRequirementsMap: Dictionary<Requirement[]> = {};
    optionalCapabilitiesMap: Dictionary<Capability[]> ={};

    constructor(@Inject(forwardRef(() => ConnectionWizardService)) public connectWizardService: ConnectionWizardService) {
    }

    ngOnInit(){
        if(this.connectWizardService.selectedMatch.isFromTo){
            this.displayRequirementsOrCapabilities = 'Capability';
            this.optionalRequirementsMap = {};
            this.optionalCapabilitiesMap = this.connectWizardService.getOptionalCapabilitiesByInstanceUniqueId(true, this.connectWizardService.selectedMatch.requirement);
        }else{
            this.displayRequirementsOrCapabilities = 'Requirement';
            this.optionalRequirementsMap = this.connectWizardService.getOptionalRequirementsByInstanceUniqueId(false, this.connectWizardService.selectedMatch.capability);
            this.optionalCapabilitiesMap = {}
        }


    }

    preventNext = ():boolean => {
        return !this.connectWizardService.selectedMatch.capability || !this.connectWizardService.selectedMatch.requirement;
    }

    preventBack = ():boolean => {
        return false;
    }

    private updateSelectedReqOrCap = (selected:Requirement|Capability):void => {
        if (!selected) {
            if (this.connectWizardService.selectedMatch.isFromTo) {
                this.connectWizardService.selectedMatch.capability = undefined;
                this.connectWizardService.selectedMatch.toNode = undefined;
            } else {
                this.connectWizardService.selectedMatch.requirement = undefined;
                this.connectWizardService.selectedMatch.fromNode = undefined;
            }
        } else if (selected instanceof Requirement) {
            this.connectWizardService.selectedMatch.requirement = <Requirement>selected;
            this.connectWizardService.selectedMatch.fromNode = this.connectWizardService.connectRelationModel.toNode.componentInstance.uniqueId;
        } else {
            this.connectWizardService.selectedMatch.capability = <Capability>selected;
            this.connectWizardService.selectedMatch.toNode = this.connectWizardService.connectRelationModel.toNode.componentInstance.uniqueId;
        }
        this.connectWizardService.selectedMatch.relationship = undefined;
    }

}
