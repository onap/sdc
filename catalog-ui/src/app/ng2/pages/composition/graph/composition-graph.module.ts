import {NgModule} from "@angular/core";
import {CommonModule} from "@angular/common";
import {CompositionGraphComponent} from "./composition-graph.component";
import {ZoneModules} from "./canvas-zone/zones-module";
import {CompositionGraphZoneUtils} from "./utils/composition-graph-zone-utils";
import {CompositionGraphGeneralUtils} from "./utils/composition-graph-general-utils";
import {CommonGraphUtils} from "./common/common-graph-utils";
import {LinksFactory} from "app/models/graph/graph-links/links-factory";
import {NodesFactory} from "app/models/graph/nodes/nodes-factory";
import {ImageCreatorService} from "./common/image-creator.service";
import {MatchCapabilitiesRequirementsUtils} from "./utils/match-capability-requierment-utils";
import {CompositionGraphNodesUtils} from "./utils/composition-graph-nodes-utils";
import {ConnectionWizardService} from "app/ng2/pages/composition/graph/connection-wizard/connection-wizard.service";
import {CompositionGraphPaletteUtils} from "./utils/composition-graph-palette-utils";
import {QueueServiceUtils} from "app/ng2/utils/queue-service-utils";
import {DndModule} from "ngx-drag-drop";
import { MenuListNg2Module } from "app/ng2/components/downgrade-wrappers/menu-list-ng2/menu-list-ng2.module";
import { UiElementsModule } from "app/ng2/components/ui/ui-elements.module";
import {ServicePathSelectorModule} from "./service-path-selector/service-path-selector.module";
import {SdcUiComponentsModule, SdcUiServices} from "onap-ui-angular";
import {CanvasSearchModule} from "./canvas-search/canvas-search.module";
import {CompositionGraphLinkUtils, ServicePathGraphUtils} from "./utils";


@NgModule({
    declarations: [CompositionGraphComponent],
    imports: [CommonModule,
        ServicePathSelectorModule,
        SdcUiComponentsModule,
        MenuListNg2Module,
        UiElementsModule,
        ZoneModules,
        CanvasSearchModule,
        DndModule],
    exports: [CompositionGraphComponent],
    entryComponents: [CompositionGraphComponent],
    providers: [
        CompositionGraphZoneUtils,
        CompositionGraphGeneralUtils,
        MatchCapabilitiesRequirementsUtils,
        CompositionGraphNodesUtils,
        CompositionGraphLinkUtils,
        CommonGraphUtils,
        NodesFactory,
        LinksFactory,
        ImageCreatorService,
        ConnectionWizardService,
        CompositionGraphPaletteUtils,
        QueueServiceUtils,
        SdcUiServices.simplePopupMenuService,
        ServicePathGraphUtils
    ]
})
export class CompositionGraphModule {
}