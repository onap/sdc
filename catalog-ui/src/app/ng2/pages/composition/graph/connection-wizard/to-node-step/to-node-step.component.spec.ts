import {async, ComponentFixture, TestBed} from "@angular/core/testing";
import {NO_ERRORS_SCHEMA} from "@angular/core";
import {ToNodeStepComponent} from "./to-node-step.component";
import {ConnectionWizardService} from "../connection-wizard.service";
import {ConfigureFn, configureTests} from "../../../../../../../jest/test-config.helper";
import {Match} from "../../../../../../models/graph/match-relation";


describe('to-node-step component', () => {

    let fixture: ComponentFixture<ToNodeStepComponent>;
    let connectionWizardServiceMock: Partial<ConnectionWizardService>;

    beforeEach(
        async(() => {

            connectionWizardServiceMock = {
                // selectedMatch: new Match(null, null, true, '',''),
                selectedMatch: {
                    isFromTo: false
                },
                getOptionalRequirementsByInstanceUniqueId: jest.fn().mockReturnValue(5),
                getOptionalCapabilitiesByInstanceUniqueId: jest.fn().mockReturnValue(10)
            }

            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [ToNodeStepComponent],
                    imports: [],
                    schemas: [NO_ERRORS_SCHEMA],
                    providers: [
                        {provide: ConnectionWizardService, useValue: connectionWizardServiceMock}
                    ],
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(ToNodeStepComponent);
            });
        })
    );


    it('should match current snapshot', () => {
        expect(fixture).toMatchSnapshot();
    });

    it('should test the ngOnInit with isFromTo = false', () => {
        const component = TestBed.createComponent(ToNodeStepComponent);
        let service = TestBed.get(ConnectionWizardService);
        service.selectedMatch.isFromTo = false;
        component.componentInstance.ngOnInit();
        expect(component.componentInstance.displayRequirementsOrCapabilities).toEqual("Requirement");
        expect(connectionWizardServiceMock.getOptionalRequirementsByInstanceUniqueId).toHaveBeenCalledWith(false, connectionWizardServiceMock.selectedMatch.capability);
        expect(component.componentInstance.optionalRequirementsMap).toEqual(5);
        expect(component.componentInstance.optionalCapabilitiesMap).toEqual({});
    });


    it('should test the ngOnInit with isFromTo = true', () => {
        const component = TestBed.createComponent(ToNodeStepComponent);
        let service = TestBed.get(ConnectionWizardService);
        service.selectedMatch.isFromTo = true;
        component.componentInstance.ngOnInit();
        expect(component.componentInstance.displayRequirementsOrCapabilities).toEqual("Capability");
        expect(connectionWizardServiceMock.getOptionalCapabilitiesByInstanceUniqueId).toHaveBeenCalledWith(true, connectionWizardServiceMock.selectedMatch.requirement);
        expect(component.componentInstance.optionalCapabilitiesMap).toEqual(10);
        expect(component.componentInstance.optionalRequirementsMap).toEqual({});
    });

});