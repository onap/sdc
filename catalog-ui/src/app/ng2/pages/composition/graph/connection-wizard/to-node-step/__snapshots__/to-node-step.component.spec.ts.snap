// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`to-node-step component should match current snapshot 1`] = `
<to-node-step
  connectWizardService={[Function Object]}
  optionalCapabilitiesMap={[Function Object]}
  optionalRequirementsMap={[Function Object]}
  preventBack={[Function Function]}
  preventNext={[Function Function]}
  updateSelectedReqOrCap={[Function Function]}
>
  <select-requirement-or-capability />
</to-node-step>
`;
